File: javascript/lesson 9/lowestCommonAncestor.js
// https://leetcode.com/problems/lowest-common-ancestor-of-a-binary-tree

function lowestCommonAncestor(root, p, q) {
    if (!root || root === p || root === q) return root;
    
    const left = lowestCommonAncestor(root.left, p, q);
    const right = lowestCommonAncestor(root.right, p, q);
    
    if (left && right) {
        return root;
    } else {
        return left || right;
    }
}

File: ruby/lesson 9/lowest_common_ancestor.rb
# https://leetcode.com/problems/lowest-common-ancestor-of-a-binary-tree

def lowest_common_ancestor(root, p, q)
    return root if root.nil? || root == p || root == q

    left = lowest_common_ancestor(root.left, p, q)
    right = lowest_common_ancestor(root.right, p, q)

    return root if !left.nil? && !right.nil?
    left || right
end
